### ISSUE TITLE
salt.roster.cache / salt.utils.cloud is_public_ip - incorrect public IP address

### ISSUE THREAD

**Issue by afletch** [saltstack/salt#39051]
`salt.utils.cloud` function `is_public_ip` is not aware of 127/8 and so when passed an address for determination on whether it is public, will return a default True when given `127.0.0.1`

`is_public_ip` is used by the salt-ssh roster `cache` in `extract_ipv4` to determine if the address is public, resulting in `127.0.0.1` being considered a public address and used for subsequent ssh connections, which are then directed at the local host.

I'm not sure if `is_public_ip` should be 127/8 aware or whether the salt-ssh cache roster should be doing additional checks, but either way it is a problem that prevents salt-ssh from working.

NB. This problem only becomes active when the ipv4 grain returns the loopback address as first in the list.

```
Salt Version:
           Salt: 2016.11.1

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.10 (default, Jul 30 2016, 19:40:32)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.2
            ZMQ: 4.2.0

System Versions:
           dist:
        machine: x86_64
        release: 16.4.0
         system: Darwin
        version: 10.12.3 x86_64
```

**Comment by gtmanfred** [comment]
Thanks for reporting this.  is_public_ip should return false for the 127/8 networks.

I am marking this as a bug.  If you would submit a pull request, we would greatly appreciate it, otherwise the riot team will handle it.

Thanks!,
Daniel

**Comment by techhat** [comment]
On it.

**Comment by techhat** [comment]
Ref #39066

**Comment by The-Loeki** [comment]
@techhat how 'bout #39050 ;)

& quite frankly I've been thinking about getting rid of that call completely; should be in salt.utils.networking or something similar IMHO

**Comment by rallytime** [comment]
@afletch This should be fixed up by #39066.
